File: cpp/TeaCrypt.cpp

/**
 * @author Rizkie Yudha Pratama <personal@rizkiepratama.net>
 * @link https://github.com/VikTymZ
 * @license MIT
 * @package TeaCrypt
 */

#include <string>
#include <iostream>
#include "TeaCrypt.h"
#include "libraries/base64.h"
#include "libraries/strrev.h"

std::string TeaCrypt::saltGenerator(int n = 5)
{
  srand(time(0));
  std::string salt;
  for(auto i = 0; i < n; i++) {
    salt += (char) rand() % 255;
  }
  return salt;
}


/**
 * Tea Encrypt
 *
 * Encryption Method
 */
std::string TeaCrypt::Encrypt(std::string data, std::string key, bool isBinarySafe = true)
{
  std::string encrypted, newKey, salt = saltGenerator();
  int cost = 1;

  // Generate new Key
  for(auto it = key.begin(), jt = salt.begin(); it != key.end(); it++ ) {
    newKey += (char) (((int) *it) ^ ((int) *jt++));
    if(jt == salt.end()) {
      jt = salt.begin();
    }
  }

  // Encrypt Our Data
  
  for(int it = 0, jt = 0, kt = 0; it < data.length(); it++) {
      // auto 
      //   iit = std::distance(data.begin(), it), 
      // ijt = std::distance(salt.begin(), jt),
      // ikt = std::distance(newKey.begin(), kt);
    // std::cout << it << " ";
    // std::cout << jt << " ";
    encrypted += (char) (
      (data[it]) ^ (newKey[jt++]) ^ (salt[kt++]) ^ (it << jt) ^ (kt >> jt) ^
      (data.length() % cost) ^ (cost >> jt) ^ (cost >> it) ^ (cost >> kt) ^ 
      (cost ^ (data.length() % (it + jt + kt + 1))) ^
      ((cost << it) % 2) ^ ((cost << jt) % 2) ^ ((cost << kt) % 2) ^
      ((cost * (it + jt + kt)) % 3)
    );
    cost++;

    if(jt == newKey.length()) {
      jt = 0;
    }

    if(kt == 5) {
      kt = 0;
    }
  }
  encrypted += salt;
  if(isBinarySafe) {
    unsigned char *t = (unsigned char *) encrypted.c_str();
    return strrev(base64_encode(t, encrypted.length()));
  } else {
    return encrypted;
  }
}



/**
 * Tea Encrypt
 *
 * Decrypt Method
 */
std::string TeaCrypt::Decrypt(std::string data, std::string key, bool isBinarySafe = true)
{
  if (isBinarySafe)
  {
    data = base64_decode(strrev(data));
  }

  std::string decrypted, newKey, salt = data.substr(data.length() - 5, data.length());
  int cost = 1;
  data = data.substr(0, data.length() - 5);

  for(auto it = key.begin(), jt = salt.begin(); it != key.end(); it++ ) {
    newKey += (char) (((int) *it) ^ ((int) *jt++));

    if( jt == salt.end() ) {
      jt = salt.begin();
    }
  }

  for(int it = 0, jt = 0, kt = 0; it < data.length(); it++) {
    // auto iit = std::distance(data.begin(), it),
    //   ijt = std::distance(salt.begin(), jt),
    //   ikt = std::distance(newKey.begin(), kt);
    // decrypted += (char) (
    // ((int) *it) ^ ((int) *jt++) ^ ((int) *kt++) ^ (iit << ijt) ^ (ikt >> ijt) ^
    // (data.length() % cost) ^ (cost >> ijt) ^ (cost >> iit) ^ (cost >> ikt) ^ 
    // (cost ^ (data.length() % (iit + ijt + ikt + 1))) ^
    // ((cost << iit) % 2) ^ ((cost << ijt) % 2) ^ ((cost << ikt) % 2) ^
    // ((cost * (iit + ijt + ikt)) % 3)
    // );
    decrypted += (char) (
      (data[it]) ^ (newKey[jt++]) ^ (salt[kt++]) ^ (it << jt) ^ (kt >> jt) ^
      (data.length() % cost) ^ (cost >> jt) ^ (cost >> it) ^ (cost >> kt) ^ 
      (cost ^ (data.length() % (it + jt + kt + 1))) ^
      ((cost << it) % 2) ^ ((cost << jt) % 2) ^ ((cost << kt) % 2) ^
      ((cost * (it + jt + kt)) % 3)
    );
    cost++;

    if(jt == newKey.length()) {
      jt = 0;
    }

    if(kt == 5) {
      kt = 0;
    }
  }

  return decrypted;
}

File: php/TeaCrypt.php
<?php

namespace TeaCrypt;

/**
 * @author Ammar Faizi <ammarfaizi2@gmail.com>
 * @link https://github.com/ammarfaizi2
 * @license MIT
 * @version 0.0.1
 * @package \TeaCrypt
 */
final class TeaCrypt
{
	/**
	 * Encrypt method.
	 *
	 * @param string $string
	 * @param string $key
	 * @param bool	 $binarySafe
	 * @return string
	 */
	final public static function encrypt($string, $key, $binarySafe = true)
	{
		$slen = strlen($string);
		$klen = strlen($key);
		$r = $newKey = "";
		$salt = self::saltGenerator();
		$cost = 1;
		for($i=$j=0;$i<$klen;$i++) {
			$newKey .= chr(ord($key[$i]) ^ ord($salt[$j++]));
			if ($j === 5) {
				$j = 0;
			}
		}	
		for($i=$j=$k=0;$i<$slen;$i++) {
			$r .= chr(
				ord($string[$i]) ^ ord($newKey[$j++]) ^ ord($salt[$k++]) ^ ($i << $j) ^ ($k >> $j) ^
				($slen % $cost) ^ ($cost >> $j) ^ ($cost >> $i) ^ ($cost >> $k) ^
				($cost ^ ($slen % ($i + $j + $k + 1))) ^ (($cost << $i) % 2) ^ (($cost << $j) % 2) ^ 
				(($cost << $k) % 2) ^ (($cost * ($i+$j+$k)) % 3)
			);
			$cost++;
			if ($j === $klen) {
				$j = 0;
			}
			if ($k === 5) {
				$k = 0;
			}
		}
		$r .= $salt;
		if ($binarySafe) {
			return strrev(base64_encode($r));
		} else {
			return $r;
		}
	}



	/**
	 * Decrypt method.
	 *
	 * @param string $string
	 * @param string $key
	 * @param bool	 $binarySafe
	 * @return string
	 */
	final public static function decrypt($string, $key, $binarySafe = true)
	{
		if ($binarySafe) {
			$string = base64_decode(strrev($string));
		}
		$slen = strlen($string);
		$salt = substr($string, $slen - 5);
		$string = substr($string, 0, ($slen = $slen - 5));
		$klen = strlen($key);
		$newKey = $r = "";
		$cost = 1;
		for($i=$j=0;$i<$klen;$i++) {
			$newKey .= chr(ord($key[$i]) ^ ord($salt[$j++]));
			if ($j === 5) {
				$j = 0;
			}
		}
		for($i=$j=$k=0;$i<$slen;$i++) {
			$r .= chr(
				ord($string[$i]) ^ ord($newKey[$j++]) ^ ord($salt[$k++]) ^ ($i << $j) ^ ($k >> $j) ^
				($slen % $cost) ^ ($cost >> $j) ^ ($cost >> $i) ^ ($cost >> $k) ^
				($cost ^ ($slen % ($i + $j + $k + 1))) ^ (($cost << $i) % 2) ^ (($cost << $j) % 2) ^ 
				(($cost << $k) % 2) ^ (($cost * ($i+$j+$k)) % 3)
			);
			$cost++;
			if ($j === $klen) {
				$j = 0;
			}
			if ($k === 5) {
				$k = 0;
			}
		}
		return $r;
	}

	/**
	 * @param int $n
	 * @return string
	 */
	final public static function saltGenerator($n = 5)
	{
		$s = range(chr(1), chr(0x7f));
		$r = ""; $c=count($s)-1;
		for($i=0;$i<$n;$i++) {
			$r.= $s[rand(0, $c)];
		}
		return $r;
	}
}